adds tools to deploy spel cicd projects with terraform

userhas404d opened the pull request:
Fixes # .

Changes offered/proposed in this pull request:
- deployment of SPEL CodeBuild projects via the cli using`pyinvoke` and terraform
- python based terraform variable assignment backed by SSM
- SPEL project SSM variable manipulation via external yaml doc


* New PR Alert to: @plus3it/spel

lorengordon commented on ci/ssm.py:
```diff
@@ -0,0 +1,253 @@
+
+def get_ssm_path(env):
+    """Return the ssm path for an environment."""
+    return "/spel/{}/".format(env)
```
This `/spel/` is hardcoded in a few places. It should be a parameter, or a global constant.

userhas404d replied:
moved all references to `/spel/` to global constant `SPEL_PATH`

lorengordon commented on ci/ssm.py:
```diff
@@ -0,0 +1,253 @@
+                if '/spel/' not in ssm_var_name:
+                    if local_variable['var_name'] == ssm_var_name:
+                        if ssm_variable['Value'] != local_variable['value']:
+                            print("Local var {} in {}".format(
```
Can't really think of a reason to separate these print statements... Combine the multiple print statements into a single print statement, and a single format statement.

userhas404d replied:
more leftovers from testing. consolidated.

lorengordon commented on ci/ssm.py:
```diff
@@ -0,0 +1,253 @@
+        for local_variable in local_values:
+            for ssm_variable in ssm_values:
+                ssm_var_name = ssm_variable['Name'].replace(ssm_path, '')
+                if '/spel/' not in ssm_var_name:
```
These nested if statements can be combined (easy to spot because the else blocks are empty)...

```
if (
    '/spel/' not in ssm_var_name and 
    local_variable['var_name'] == ssm_var_name
):
```

userhas404d replied:
consolidated

lorengordon commented on ci/ssm.py:
```diff
@@ -0,0 +1,253 @@
+                        if local_variable['var_name'] == ssm_var_name:
+                            if (
+                              ssm_variable['Value'] != local_variable['value']
+                               ):
```
Align this line with the if statement

userhas404d replied:
fixed

lorengordon commented on ci/ssm.py:
```diff
@@ -0,0 +1,253 @@
+            for local_variable in local_values[env]:
+                for ssm_variable in ssm_values:
+                    ssm_var_name = ssm_variable['Name'].replace(ssm_path, '')
+                    if '/spel/' not in ssm_var_name:
```
Combine these if statements

userhas404d replied:
combined

lorengordon commented on ci/ssm.py:
```diff
@@ -0,0 +1,253 @@
+def get_env_paths():
+    """Get environment variable (ssm) paths."""
+    environment_paths = []
+    zones = [
```
this is defined in tasks.py. import it as a global constant.

userhas404d replied:
imported

lorengordon commented on ci/ssm.py:
```diff
@@ -0,0 +1,253 @@
+        "commercial",
+        "govcloud"
+    ]
+    project_types = [
```
this is defined in tasks.py. import it as a global constant.

userhas404d replied:
imported

userhas404d commented on ci/modules/amis/main.tf:
```diff
@@ -0,0 +1,117 @@
+  value = [
+    {
+      "name"  = "SOURCE_AMI_CENTOS6_HVM"
+      "value" = "${data.aws_ami.centos6_hvm_com.*.id[0]}"
```
fixed

userhas404d commented on ci/modules/iam/main.tf:
```diff
@@ -0,0 +1,154 @@
+  name = "packer-codebuild-permissions"
+  role = "${aws_iam_role.spel-codebuild-role.name}"
+
+  policy = <<-POLICY
```
moved to inline json

userhas404d commented on ci/modules/iam/variables.tf:
```diff
@@ -0,0 +1,14 @@
+
+variable "ssm_key_name" {
+  type        = "string"
+  description = "Ailas of the target account's ssm key"
```
fixed

userhas404d commented on ci/modules/remote_state/main.tf:
```diff
@@ -0,0 +1,77 @@
+  }
+}
+
+resource "null_resource" "delete_s3_objects" {
```
This was me looping back to `pyinvoke` to call `s3.py` and delete the versioned tfstate bucket(s). Should probably be an external data source.

userhas404d commented on ci/modules/remote_state/outputs.tf:
```diff
@@ -0,0 +1,9 @@
+}
+
+output "dynamodb_table_name" {
+  value       = "${join("", aws_dynamodb_table.default.*.name)}"
```
fixed

userhas404d commented on ci/modules/variables.tf:
```diff
@@ -0,0 +1,50 @@
+  type = "list"
+
+  default = [
+    {
```
updated

userhas404d commented on ci/modules/variables.tf:
```diff
@@ -0,0 +1,50 @@
+  description = "What the codebuild project is tarteting ie: commercial_ci, govcloud_rc, etc."
+}
+
+variable "root_bucket" {
```
removed

userhas404d commented on ci/modules/iam/main.tf:
```diff
@@ -0,0 +1,154 @@
+            {
+                "Effect": "Allow",
+                "Resource": [
+                    "arn:aws:logs:us-east-1:${var.aws_account_id}:log-group:/aws/codebuild/spel-ci",
```
updated

userhas404d commented on ci/ssm.py:
```diff
@@ -0,0 +1,253 @@
+                        print('A variable block in the {} section was found'
+                              ' with a blank a var_name'
+                              ' declaration'.format(key))
+                except KeyError as e:
```
fixed. Seeing these now with vs code so should (hopefully) be a non issue in the future.

userhas404d commented on ci/ssm.py:
```diff
@@ -0,0 +1,253 @@
+        param_obj = yaml.load(stream)
+        return param_obj
+    except FileNotFoundError:
+        print('variables.yaml file was not found. Make sure it exists and is' +
```
fixed.

userhas404d commented on ci/ssm.py:
```diff
@@ -0,0 +1,253 @@
+                try:
+                    if var['var_type'] != "SecureString"  \
+                            and var['var_type'] != "String":
+                            print('\nA variable block in the {} section was'
```
No reason in particular. Guessing it was there originally to get that statement to stand out while testing. removed.

userhas404d commented on ci/modules/remote_state/main.tf:
```diff
@@ -0,0 +1,77 @@
+  count = "${var.automate_build ? 1 : 0}"
+
+  triggers = {
+    uuid_tirgger        = "${uuid()}"                                       # https://github.com/hashicorp/terraform/issues/7603#issuecomment-232085366
```
fixed

userhas404d commented on ci/modules/remote_state/main.tf:
```diff
@@ -0,0 +1,77 @@
+
+  triggers = {
+    uuid_tirgger        = "${uuid()}"                                       # https://github.com/hashicorp/terraform/issues/7603#issuecomment-232085366
+    s3_bucket_id        = "${join(" ", aws_s3_bucket.default.*.id)}"
```
updated

userhas404d commented on ci/modules/remote_state/main.tf:
```diff
@@ -0,0 +1,77 @@
+  triggers = {
+    uuid_tirgger        = "${uuid()}"                                       # https://github.com/hashicorp/terraform/issues/7603#issuecomment-232085366
+    s3_bucket_id        = "${join(" ", aws_s3_bucket.default.*.id)}"
+    dynamodb_table_name = "${join(" ", aws_dynamodb_table.default.*.name)}"
```
updated

userhas404d commented:
I reworked the backend creation module to use cloudposse's and refactored `tasks.py` to hopefully be a bit more succinct/readable. Outputs of the backend module (bucket name and DB name) are now being captured in SSM via `tasks.py` and passed on `terraform init` of the codebuild module.
